File: packages/backend/src/routes/authRoutes.ts
/**
 * 认证相关路由
 * 包含登录、注册、登出、刷新token等功能
 */

import { Router, Request, Response } from 'express';
import { AuthService } from '../services/AuthService';
import { prisma } from '../db/prisma';
import { authenticate } from '../middleware/authMiddleware';

const router = Router();
const authService = new AuthService(prisma);

/**
 * POST /api/auth/register
 * 用户注册
 */
router.post('/register', async (req: Request, res: Response) => {
  try {
    const { username, email, password, displayName } = req.body;

    // 验证必填字段
    if (!username || !email || !password) {
      res.status(400).json({
        success: false,
        error: 'Missing required fields',
        message: '缺少必填字段'
      });
      return;
    }

    // 注册用户
    const result = await authService.register({
      username,
      email,
      password,
      displayName
    });

    res.status(201).json({
      success: true,
      data: result,
      message: '注册成功'
    });
  } catch (error: any) {
    console.error('Register error:', error);
    res.status(400).json({
      success: false,
      error: error.message,
      message: '注册失败'
    });
  }
});

/**
 * POST /api/auth/login
 * 用户登录
 */
router.post('/login', async (req: Request, res: Response) => {
  try {
    const { email, password } = req.body;

    // 验证必填字段
    if (!email || !password) {
      res.status(400).json({
        success: false,
        error: 'Missing required fields',
        message: '缺少邮箱或密码'
      });
      return;
    }

    // 登录
    const result = await authService.login({ email, password });

    res.json({
      success: true,
      data: result,
      message: '登录成功'
    });
  } catch (error: any) {
    console.error('Login error:', error);
    res.status(401).json({
      success: false,
      error: error.message,
      message: '登录失败'
    });
  }
});

/**
 * POST /api/auth/logout
 * 用户登出
 */
router.post('/logout', authenticate, async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];
    
    if (token) {
      await authService.logout(token);
    }

    res.json({
      success: true,
      message: '登出成功'
    });
  } catch (error: any) {
    console.error('Logout error:', error);
    res.status(500).json({
      success: false,
      error: error.message,
      message: '登出失败'
    });
  }
});

/**
 * POST /api/auth/refresh
 * 刷新Token
 */
router.post('/refresh', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];

    if (!token) {
      res.status(401).json({
        success: false,
        error: 'No token provided',
        message: '缺少Token'
      });
      return;
    }

    const result = await authService.refreshToken(token);

    res.json({
      success: true,
      data: result,
      message: 'Token刷新成功'
    });
  } catch (error: any) {
    console.error('Refresh token error:', error);
    res.status(401).json({
      success: false,
      error: error.message,
      message: 'Token刷新失败'
    });
  }
});

/**
 * GET /api/auth/me
 * 获取当前用户信息
 */
router.get('/me', authenticate, async (req: Request, res: Response) => {
  try {
    res.json({
      success: true,
      data: req.user,
      message: '获取用户信息成功'
    });
  } catch (error: any) {
    console.error('Get current user error:', error);
    res.status(500).json({
      success: false,
      error: error.message,
      message: '获取用户信息失败'
    });
  }
});

/**
 * POST /api/auth/change-password
 * 修改密码
 */
router.post('/change-password', authenticate, async (req: Request, res: Response) => {
  try {
    const { oldPassword, newPassword } = req.body;

    if (!oldPassword || !newPassword) {
      res.status(400).json({
        success: false,
        error: 'Missing required fields',
        message: '缺少旧密码或新密码'
      });
      return;
    }

    await authService.changePassword(
      req.user!.id,
      oldPassword,
      newPassword
    );

    res.json({
      success: true,
      message: '密码修改成功，请重新登录'
    });
  } catch (error: any) {
    console.error('Change password error:', error);
    res.status(400).json({
      success: false,
      error: error.message,
      message: '密码修改失败'
    });
  }
});

/**
 * GET /api/auth/check
 * 检查token是否有效
 */
router.get('/check', async (req: Request, res: Response) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];

    if (!token) {
      res.json({
        success: true,
        data: { valid: false },
        message: '未提供Token'
      });
      return;
    }

    const payload = await authService.verifyToken(token);

    res.json({
      success: true,
      data: {
        valid: !!payload,
        payload: payload || null
      },
      message: payload ? 'Token有效' : 'Token无效'
    });
  } catch (error: any) {
    res.json({
      success: true,
      data: { valid: false },
      message: 'Token无效'
    });
  }
});

export default router;

File: backend/app/api/auth_routes.py
"""
Authentication Routes - JWT token-based authentication
Implements login, logout, token refresh, and user management
"""

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from sqlalchemy.orm import Session
from datetime import timedelta, datetime
from typing import Optional

from app.models.user import (
    User, UserCreate, UserResponse, UserUpdate, Token, TokenData
)
from app.core.user_database import (
    get_db, authenticate_user, create_user, get_user_by_username,
    get_user_by_id, update_user, delete_user
)
from app.core.security import (
    create_access_token, create_refresh_token, decode_access_token,
    ACCESS_TOKEN_EXPIRE_MINUTES
)
from app.core.login_attempts import get_login_tracker
from app.core.logger import get_logger

logger = get_logger(__name__)
router = APIRouter(prefix="/auth", tags=["authentication"])

# OAuth2 password bearer for token extraction
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/auth/login")


async def get_current_user(
    token: str = Depends(oauth2_scheme),
    db: Session = Depends(get_db)
) -> User:
    """
    Dependency to get current authenticated user from JWT token.
    
    Args:
        token: JWT token from Authorization header
        db: Database session
        
    Returns:
        Current authenticated User object
        
    Raises:
        HTTPException: If token is invalid or user not found
    """
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    
    # Decode token
    payload = decode_access_token(token)
    
    if payload is None:
        raise credentials_exception
    
    # Extract username from token
    username: str = payload.get("sub")
    
    if username is None:
        raise credentials_exception
    
    # Get user from database
    user = get_user_by_username(db, username=username)
    
    if user is None:
        raise credentials_exception
    
    if not user.is_active:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Inactive user account"
        )
    
    return user


async def get_current_active_superuser(
    current_user: User = Depends(get_current_user)
) -> User:
    """
    Dependency to get current superuser (admin).
    Use this to protect admin-only routes.
    """
    if not current_user.is_superuser:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    return current_user


from app.services.auth_service import AuthService

# ... imports ...

@router.post("/login", response_model=Token)
async def login(
    form_data: OAuth2PasswordRequestForm = Depends(),
    db: Session = Depends(get_db)
):
    """
    OAuth2 compatible token login.
    Delegates logic to AuthService.
    """
    auth_service = AuthService(db)
    return auth_service.login_user(form_data.username, form_data.password)


@router.post("/register", response_model=UserResponse)
async def register(
    user: UserCreate,
    db: Session = Depends(get_db)
):
    """
    Register a new user account.
    
    Request body:
        - username: Unique username (3-50 chars, alphanumeric, -, _)
        - email: Valid email address
        - password: Password (min 8 chars)
        - full_name: Optional full name
        
    Returns:
        Created user object (without password)
    """
    try:
        db_user = create_user(db, user)
        logger.info("user_registered",
                   user_id=db_user.id,
                   username=db_user.username,
                   email=db_user.email)
        return db_user
    except ValueError as e:
        logger.warning("registration_failed",
                      username=user.username,
                      error=str(e))
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/me", response_model=UserResponse)
async def read_users_me(
    current_user: User = Depends(get_current_user)
):
    """
    Get current user information.
    Requires valid JWT token in Authorization header.
    """
    return current_user


@router.put("/me", response_model=UserResponse)
async def update_user_me(
    user_update: UserUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Update current user's information.
    Users can update their own email, full_name, and password.
    """
    updated_user = update_user(db, current_user.id, user_update)
    
    if not updated_user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    return updated_user


@router.post("/refresh", response_model=Token)
async def refresh_token(
    current_user: User = Depends(get_current_user)
):
    """
    Refresh access token using existing valid token.
    Returns a new access token with extended expiration.
    """
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": current_user.username, "user_id": current_user.id},
        expires_delta=access_token_expires
    )
    
    return {
        "access_token": access_token,
        "token_type": "bearer",
        "expires_in": ACCESS_TOKEN_EXPIRE_MINUTES * 60
    }


@router.post("/logout")
async def logout(current_user: User = Depends(get_current_user)):
    """
    Logout current user.
    Note: JWT tokens are stateless, so actual logout is handled client-side
    by deleting the stored token. This endpoint is for logging/auditing purposes.
    """
    logger.info("user_logged_out",
               user_id=current_user.id,
               username=current_user.username)
    return {
        "message": "Successfully logged out",
        "username": current_user.username
    }


# Admin routes
@router.get("/users", response_model=list[UserResponse])
async def list_users(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_active_superuser),
    db: Session = Depends(get_db)
):
    """
    List all users (admin only).
    """
    users = db.query(User).offset(skip).limit(limit).all()
    return users


@router.get("/users/{user_id}", response_model=UserResponse)
async def get_user(
    user_id: str,
    current_user: User = Depends(get_current_active_superuser),
    db: Session = Depends(get_db)
):
    """
    Get user by ID (admin only).
    """
    user = get_user_by_id(db, user_id)
    
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    return user


@router.put("/users/{user_id}", response_model=UserResponse)
async def admin_update_user(
    user_id: str,
    user_update: UserUpdate,
    current_user: User = Depends(get_current_active_superuser),
    db: Session = Depends(get_db)
):
    """
    Update any user (admin only).
    Admins can update is_active flag.
    """
    updated_user = update_user(db, user_id, user_update)
    
    if not updated_user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    return updated_user


@router.delete("/users/{user_id}")
async def admin_delete_user(
    user_id: str,
    current_user: User = Depends(get_current_active_superuser),
    db: Session = Depends(get_db)
):
    """
    Delete user (admin only).
    """
    success = delete_user(db, user_id)
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    return {"message": "User deleted successfully"}
